src/index.ts

```diff
@@ -21,6 +21,18 @@ type Requires<Schemes extends BaseSchemes> =
   | RenderSignal<'connection', { payload: Schemes['Connection'], start?: Position, end?: Position }>
   | { type: 'unmount', data: { element: HTMLElement } }
 
+/**
+ * Vue plugin options used to setup vue instance(s) used by retejs.
+ */
+export type Props = {
+  /**
+   * Use this to setup vue.
+   *  @param [context] to be used for createApp({ ...context }) or new Vue({ ...context })
+   *  @returns app / vue instance.
+   */
+  setup?: (context: object) => object;
```

Comment:
```ts
  setup?: (context: Context) => Instance;
```

these types can be defined in `./vuecompat/types.ts`:

```ts

import type VueNamespace from 'vue'

type Vue = typeof VueNamespace

export type Context = Vue extends { createApp: (arg: infer U) => any }
  ? U
  : (Vue extends { new(options: infer U): any } ? U : any)

export type Instance = Vue extends { createApp: (arg: any) => infer U }
  ? U
  : (Vue extends { new(options: any): infer U } ? U : any)
```

This way, it can infer an options and return type depending on Vue.js version

Reply:
i squashed my commits to one "feat" commit for the new configuration feature.
i also added your dynamic typings as "refactor" commit, looks fine for me.